File: bms-tool-android/app/src/main/java/com/cjsh/bmsassistant/protocol/ModbusRtu.kt
package com.cjsh.bmsassistant.protocol

import java.io.IOException
import java.nio.charset.StandardCharsets

object BmsRegisters {
    const val DEVICE_ADDRESS = 0x01
    const val MAC = 0x0000
    const val BT_NAME = 0x0100
    const val SERIAL = 0xC002
    const val HARDWARE = 0xC012
    const val SOFTWARE = 0xC022
    const val EVENT_LOG = 0xC008
    const val PROTECT = 0x2100
    const val PROTECT_COUNT = 65
    const val AFE = 0x2400
    const val AFE_COUNT = 24
    const val LEGACY = 0xD000
    const val SYSTEM_STATUS = 0xD115
    const val REALTIME = 0xD120
    const val REALTIME_MAGIC = 0x4253
    const val BT_NAME_READ_WORDS = 12
    const val BT_NAME_MAX_SUFFIX_BYTES = 10
}

object ModbusRtu {
    fun readHolding(start: Int, quantity: Int): ByteArray {
        require(quantity in 1..125)
        return frame(byteArrayOf(
            BmsRegisters.DEVICE_ADDRESS.toByte(), 0x03,
            (start ushr 8).toByte(), start.toByte(),
            (quantity ushr 8).toByte(), quantity.toByte()
        ))
    }

    fun writeSingle(register: Int, value: Int): ByteArray = frame(byteArrayOf(
        BmsRegisters.DEVICE_ADDRESS.toByte(), 0x06,
        (register ushr 8).toByte(), register.toByte(),
        (value ushr 8).toByte(), value.toByte()
    ))

    fun writeMultiple(start: Int, values: IntArray): ByteArray {
        require(values.isNotEmpty())
        val raw = ByteArray(values.size * 2)
        values.forEachIndexed { i, v ->
            require(v in 0..0xFFFF)
            raw[i * 2] = (v ushr 8).toByte()
            raw[i * 2 + 1] = v.toByte()
        }
        return writeMultipleRaw(start, raw)
    }

    fun writeMultipleRaw(start: Int, raw: ByteArray): ByteArray {
        require(raw.isNotEmpty() && raw.size % 2 == 0)
        val qty = raw.size / 2
        val body = ByteArray(7 + raw.size)
        body[0] = BmsRegisters.DEVICE_ADDRESS.toByte()
        body[1] = 0x10
        body[2] = (start ushr 8).toByte(); body[3] = start.toByte()
        body[4] = (qty ushr 8).toByte(); body[5] = qty.toByte()
        body[6] = raw.size.toByte()
        raw.copyInto(body, 7)
        return frame(body)
    }

    fun parseRead(frame: ByteArray, expectedQuantity: Int): IntArray {
        validateFrame(frame)
        val function = frame[1].toInt() and 0xFF
        if ((function and 0x80) != 0) throw IOException("Modbus exception code=0x%02X".format(frame[2].toInt() and 0xFF))
        if ((frame[0].toInt() and 0xFF) != BmsRegisters.DEVICE_ADDRESS || function != 0x03) throw IOException("Unexpected Modbus read response")
        val bytes = frame[2].toInt() and 0xFF
        if (bytes != expectedQuantity * 2 || frame.size != bytes + 5) throw IOException("Modbus read response length mismatch")
        return IntArray(expectedQuantity) { i ->
            ((frame[3 + i * 2].toInt() and 0xFF) shl 8) or (frame[4 + i * 2].toInt() and 0xFF)
        }
    }

    fun validateWriteSingleAck(frame: ByteArray, register: Int, value: Int) {
        validateFrame(frame)
        val f = frame[1].toInt() and 0xFF
        if ((f and 0x80) != 0) throw IOException("Modbus exception code=0x%02X".format(frame[2].toInt() and 0xFF))
        if (frame.size != 8 || f != 0x06 || u16(frame, 2) != register || u16(frame, 4) != value) throw IOException("Single-register write acknowledgement mismatch")
    }

    fun validateWriteMultipleAck(frame: ByteArray, register: Int, quantity: Int) {
        validateFrame(frame)
        val f = frame[1].toInt() and 0xFF
        if ((f and 0x80) != 0) throw IOException("Modbus exception code=0x%02X".format(frame[2].toInt() and 0xFF))
        if (frame.size != 8 || f != 0x10 || u16(frame, 2) != register || u16(frame, 4) != quantity) throw IOException("Multi-register write acknowledgement mismatch")
    }

    fun inferExpectedLength(buffer: List<Byte>): Int? {
        if (buffer.size < 2) return null
        val f = buffer[1].toInt() and 0xFF
        if ((f and 0x80) != 0) return 5
        return when (f) {
            0x03 -> if (buffer.size >= 3) (buffer[2].toInt() and 0xFF) + 5 else null
            0x06, 0x10 -> 8
            else -> null
        }
    }

    fun validateFrame(frame: ByteArray) {
        if (frame.size < 4) throw IOException("Modbus response too short")
        val rx = (frame[frame.size - 2].toInt() and 0xFF) or ((frame.last().toInt() and 0xFF) shl 8)
        val calc = crc16(frame, 0, frame.size - 2)
        if (rx != calc) throw IOException("Modbus CRC mismatch: rx=0x%04X calc=0x%04X".format(rx, calc))
    }

    fun frame(body: ByteArray): ByteArray {
        val out = body.copyOf(body.size + 2)
        val crc = crc16(body, 0, body.size)
        out[out.size - 2] = crc.toByte()
        out[out.size - 1] = (crc ushr 8).toByte()
        return out
    }

    fun crc16(data: ByteArray, offset: Int = 0, length: Int = data.size - offset): Int {
        var crc = 0xFFFF
        for (i in offset until offset + length) {
            crc = crc xor (data[i].toInt() and 0xFF)
            repeat(8) { crc = if ((crc and 1) != 0) (crc ushr 1) xor 0xA001 else crc ushr 1 }
        }
        return crc and 0xFFFF
    }

    fun decodeAscii(words: IntArray): String {
        val bytes = ArrayList<Byte>(words.size * 2)
        words.forEach { w -> bytes.add((w ushr 8).toByte()); bytes.add(w.toByte()) }
        val raw = bytes.takeWhile { it.toInt() != 0 }.toByteArray()
        return String(raw, StandardCharsets.UTF_8).trim()
    }

    fun hex(data: ByteArray): String = data.joinToString("") { "%02X".format(it.toInt() and 0xFF) }
    private fun u16(data: ByteArray, off: Int) = ((data[off].toInt() and 0xFF) shl 8) or (data[off + 1].toInt() and 0xFF)
}

File: bms-tool-windows/BmsTool.Windows/ModbusRtu.cs
using System.Buffers.Binary;
using System.IO;
using System.Text;

namespace BmsTool.Windows;

public static class BmsRegisters
{
    public const byte DeviceAddress = 0x01;
    public const ushort Mac = 0x0000;
    public const ushort BtName = 0x0100;
    public const ushort Serial = 0xC002;
    public const ushort Hardware = 0xC012;
    public const ushort Software = 0xC022;
    public const ushort Protect = 0x2100;
    public const ushort Legacy = 0xD000;
    public const ushort SystemStatus = 0xD115;
    public const ushort Realtime = 0xD120;
    public const ushort RealtimeMagic = 0x4253;
    public const int SeriesCount = 10;
    public const int BtNameReadWords = 12;
    public const int BtNameMaxSuffixBytesPerBleRequest = 10;
}

public static class ModbusRtu
{
    public static byte[] ReadHolding(ushort start, ushort quantity)
    {
        Span<byte> body = stackalloc byte[6];
        body[0] = BmsRegisters.DeviceAddress; body[1] = 0x03;
        BinaryPrimitives.WriteUInt16BigEndian(body[2..4], start);
        BinaryPrimitives.WriteUInt16BigEndian(body[4..6], quantity);
        return Frame(body);
    }

    public static byte[] WriteMultiple(ushort start, ReadOnlySpan<byte> rawRegisterBytes)
    {
        if (rawRegisterBytes.Length == 0 || (rawRegisterBytes.Length & 1) != 0) throw new ArgumentException("Register payload must contain whole words.");
        ushort qty = checked((ushort)(rawRegisterBytes.Length / 2));
        byte[] body = new byte[7 + rawRegisterBytes.Length];
        body[0] = BmsRegisters.DeviceAddress; body[1] = 0x10;
        BinaryPrimitives.WriteUInt16BigEndian(body.AsSpan(2, 2), start);
        BinaryPrimitives.WriteUInt16BigEndian(body.AsSpan(4, 2), qty);
        body[6] = checked((byte)rawRegisterBytes.Length);
        rawRegisterBytes.CopyTo(body.AsSpan(7));
        return Frame(body);
    }

    public static ushort[] ParseRead(byte[] frame, ushort expectedQuantity)
    {
        ValidateFrame(frame);
        if (frame[0] != BmsRegisters.DeviceAddress) throw new IOException("Unexpected Modbus slave address.");
        if ((frame[1] & 0x80) != 0) throw new IOException($"Modbus exception function=0x{frame[1] & 0x7F:X2}, code=0x{frame[2]:X2}.");
        if (frame[1] != 0x03) throw new IOException($"Expected function 0x03, got 0x{frame[1]:X2}.");
        int bytes = frame[2];
        if (bytes != expectedQuantity * 2 || frame.Length != bytes + 5) throw new IOException("Modbus read response length mismatch.");
        ushort[] words = new ushort[expectedQuantity];
        for (int i = 0; i < words.Length; i++) words[i] = BinaryPrimitives.ReadUInt16BigEndian(frame.AsSpan(3 + i * 2, 2));
        return words;
    }

    public static void ValidateWriteMultipleAck(byte[] frame, ushort expectedRegister, ushort expectedQuantity)
    {
        ValidateFrame(frame);
        if ((frame[1] & 0x80) != 0) throw new IOException($"Modbus exception code=0x{frame[2]:X2}.");
        if (frame[0] != BmsRegisters.DeviceAddress || frame[1] != 0x10 || frame.Length != 8) throw new IOException("Invalid write acknowledgement.");
        if (BinaryPrimitives.ReadUInt16BigEndian(frame.AsSpan(2,2)) != expectedRegister || BinaryPrimitives.ReadUInt16BigEndian(frame.AsSpan(4,2)) != expectedQuantity) throw new IOException("Write acknowledgement mismatch.");
    }

    public static int? InferExpectedLength(IReadOnlyList<byte> buffer)
    {
        if (buffer.Count < 2) return null;
        byte f = buffer[1];
        if ((f & 0x80) != 0) return 5;
        if (f == 0x03) return buffer.Count >= 3 ? buffer[2] + 5 : null;
        if (f == 0x06 || f == 0x10) return 8;
        return null;
    }

    public static byte[] Frame(ReadOnlySpan<byte> body)
    {
        byte[] result = new byte[body.Length + 2]; body.CopyTo(result);
        ushort crc = Crc16(body); result[^2] = (byte)crc; result[^1] = (byte)(crc >> 8); return result;
    }

    public static void ValidateFrame(ReadOnlySpan<byte> frame)
    {
        if (frame.Length < 4) throw new IOException("Modbus response too short.");
        ushort rx = (ushort)(frame[^2] | frame[^1] << 8); ushort calc = Crc16(frame[..^2]);
        if (rx != calc) throw new IOException($"Modbus CRC mismatch: rx=0x{rx:X4}, calc=0x{calc:X4}.");
    }

    public static ushort Crc16(ReadOnlySpan<byte> data)
    {
        ushort crc = 0xFFFF;
        foreach (byte b in data) { crc ^= b; for (int i=0;i<8;i++) crc = (crc & 1) != 0 ? (ushort)((crc >> 1) ^ 0xA001) : (ushort)(crc >> 1); }
        return crc;
    }

    public static string DecodeAscii(IEnumerable<ushort> words)
    {
        var bytes = words.SelectMany(w => new[]{(byte)(w>>8),(byte)w}).ToList(); int zero = bytes.IndexOf(0); if (zero >= 0) bytes.RemoveRange(zero, bytes.Count-zero); return Encoding.UTF8.GetString(bytes.ToArray()).Trim();
    }
}
